### PR TITLE
Make DSP code dumping consistent between DSP-LLE and DSP-HLE

### PR DESCRIPTION
**Pull request by ligfx**
- DSP-LLE dumps code when `m_DumpUCode` is set instead of only/always in debug builds.
- De-duplicated dumping/file writing logic that was in three different places.
- DSP-HLE automatically disassembles dumped code and writes it to an accompanying
text file.
- Makes DSP-LLE code checksums the same as those from DSP-HLE. I'm assuming DSP-HLE was doing it correctly, since there are numerous references to the pre-endian-swapped checksums.
- Fixes disassembly when dumping code from DSP-LLE, which was using the wrong endianness and giving totally bogus output.
- Reveals error messages of the format
  > "Bah! ReadAnnotatedAssembly couldn't find the file ../../docs/DSP/DSP_UC_AX_07F88145.txt" 
  
  which seems to be intended behavior that was previously hidden.
- Couple other code cleanups while I'm here.

### CODE REVIEW

**Comment by ligfx** [comment]
Okay, now the Windows build of DSPTool is raising a _ton_ of link errors. I haven't had the time to look at the DSPTool code much, anybody have ideas on the best way to go about getting the build to work?

File: Source/DSPTool/DSPTool.vcxproj [hunk 1]
```diff
@@ -37,6 +37,7 @@
   <ItemDefinitionGroup>
     <Link>
       <AdditionalDependencies>winmm.lib;%(AdditionalDependencies)</AdditionalDependencies>
+      <AdditionalLibraryDirectories>$(ExternalsDir)OpenAL\lib;$(ExternalsDir)ffmpeg\lib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
```
Comment (Sarkie) [inline Source/DSPTool/DSPTool.vcxproj]: winmm.lib; is now missing?

File: Source/Core/Core/DSP/DSPCodeUtil.cpp [hunk 2]
```diff
@@ -218,4 +218,44 @@ bool SaveBinary(const std::vector<u16>& code, const std::string& filename)
     return false;
   return true;
 }
+
+bool DumpDSPCode(const u8* code_be, int size_in_bytes, u32 crc)
+{
+  const std::string binFile =
+      StringFromFormat("%sDSP_UC_%08X.bin", File::GetUserPath(D_DUMPDSP_IDX).c_str(), crc);
+  const std::string txtFile =
```
Comment (BhaaLseN) [inline Source/Core/Core/DSP/DSPCodeUtil.cpp]: Code style: `bin_file` and `txt_file`. Might as well fix that when you're touching the code anyways.

File: Source/Core/Core/DSP/DSPCodeUtil.cpp [hunk 3]
```diff
@@ -218,4 +218,44 @@ bool SaveBinary(const std::vector<u16>& code, const std::string& filename)
...
+      StringFromFormat("%sDSP_UC_%08X.txt", File::GetUserPath(D_DUMPDSP_IDX).c_str(), crc);
+
+  File::IOFile pFile(binFile, "wb");
```
Comment (BhaaLseN) [inline Source/Core/Core/DSP/DSPCodeUtil.cpp]: Same here, whats wrong with a simple `file`?

File: Source/Core/Core/DSP/DSPHWInterface.cpp [hunk 4]
```diff
@@ -227,20 +230,24 @@ u16 gdsp_ifx_read(u16 addr)
...
+    DumpDSPCode(code, size, g_dsp.iram_crc);
+
   Common::UnWriteProtectMemory(g_dsp.iram, DSP_IRAM_BYTE_SIZE, false);
 
   u8* dst = ((u8*)g_dsp.iram);
   for (u32 i = 0; i < size; i += 2)
   {
-    *(u16*)&dst[dsp_addr + i] =
-        Common::swap16(*(const u16*)&g_dsp.cpu_ram[(addr + i) & 0x0fffffff]);
+    *(u16*)&dst[dsp_addr + i] = Common::swap16(*(const u16*)&code[i]);
```
Comment (BhaaLseN) [inline Source/Core/Core/DSP/DSPHWInterface.cpp]: This is kinda undefined behavior; this should probably use `memcpy`/`memmove` instead (technically, both before and after the equal sign). You can also look at the overload of `Common::swap16` that takes a pointer (vs. the one you're calling right now that takes a `u16`).
Should be fixed at some point, and might as well do it now that you're touching the code anyways.

File: Source/Core/Core/DSP/DSPCodeUtil.cpp (hunk 2 above)
Comment (ligfx) [inline Source/Core/Core/DSP/DSPCodeUtil.cpp]: Done. This function had a bunch of spots to clean up, I did the others as well.

File: Source/Core/Core/DSP/DSPCodeUtil.cpp (hunk 3 above)
Comment (ligfx) [inline Source/Core/Core/DSP/DSPCodeUtil.cpp]: "This file has been brought to you by the letter 'P'."

File: Source/Core/Core/DSP/DSPHWInterface.cpp (hunk 4 above)
Comment (ligfx) [inline Source/Core/Core/DSP/DSPHWInterface.cpp]: Oog, this is a mess. Yeah, fixed.

**Comment by ligfx** [comment]
@skidau you were right, moving stuff around with the Host_* functions fixed the Windows build issue. Thanks for the help!

File: Source/Core/Core/DSP/DSPHWInterface.cpp [hunk 5]
```diff
@@ -227,22 +228,22 @@ u16 gdsp_ifx_read(u16 addr)
...
-    *(u16*)&dst[dsp_addr + i] =
-        Common::swap16(*(const u16*)&g_dsp.cpu_ram[(addr + i) & 0x0fffffff]);
-  }
-  Common::WriteProtectMemory(g_dsp.iram, DSP_IRAM_BYTE_SIZE, false);
+  const u8* code = &g_dsp.cpu_ram[addr & 0x0fffffff];
+  g_dsp.iram_crc = HashEctor(code, size);
 
-  Host::CodeLoaded((const u8*)g_dsp.iram + dsp_addr, size);
+  Common::UnWriteProtectMemory(g_dsp.iram, DSP_IRAM_BYTE_SIZE, false);
+  memcpy(dst, code, size);
```
Comment (lioncash) [inline Source/Core/Core/DSP/DSPHWInterface.cpp]: `<cstddef>` and `<cstring>` need to be included.

Comment (ligfx) [inline Source/Core/Core/DSP/DSPHWInterface.cpp]: Good eyes!

**Comment by linxbek** [comment]
Can someone smart add support of vulkan on mali g71 gpu ? This gpu actually does have vulkan but none version pic dolphin can run it

**Pull request merged**